README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> NATS messaging components for Pip.Services in Go

This library is a part of [Pip.Services](https://github.com/pip-services/pip-services) project.
The Nats module contains a set of components for working with the message queue via NATS server [https://nats.io/](https://nats.io/).

The module contains the following packages:
- **Build** - Factory for constructing module components
- **Connect** - Components for creating and configuring a connection with Nats
- **Queues** - Message Queuing components that implement the standard [Messaging](https://github.com/pip-services3-gox/pip-services3-messaging-gox) module interface

<a name="links"></a> Quick links:

* [Configuration](http://docs.pipservices.org/concepts/configuration/)
* [API Reference](https://godoc.org/github.com/pip-services3-gox/pip-services3-nats-gox/)
* [Change Log](CHANGELOG.md)
* [Get Help](http://docs.pipservices.org/get_help/)
* [Contribute](http://docs.pipservices.org/contribute/)

## Use

Get the package from the Github repository:
```bash
go get -u github.com/pip-services3-gox/pip-services3-nats-gox@latest
```

## Develop

For development you shall install the following prerequisites:
* Golang v1.18+
* Visual Studio Code or another IDE of your choice
* Docker
* Git

Run automated tests:
```bash
go test -v ./test/...
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized test as:
```bash
./test.ps1
./clear.ps1
```

## Contacts

The Golang version of Pip.Services is created and maintained by:
- **Sergey Seroukhov**

The documentation is written by:
- **Levichev Dmitry**


CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Nats specific components for Golang Changelog

## <a name="1.0.1"></a> 1.0.1 (2022-07-10)

- Updated dependencies

## <a name="1.0.0"></a> 1.0.0 (2022-07-10)

### Features
* **build** Factory for constructing module components
* **connect** Components for creating and configuring a connection
* **queues** Message Queuing components

test/queues/MessageQueueFixture.go
package test_queues

import (
	"context"
	"sync"
	"testing"
	"time"

	"github.com/pip-services3-gox/pip-services3-messaging-gox/queues"
	"github.com/stretchr/testify/assert"
)

type MessageQueueFixture struct {
	queue queues.IMessageQueue
}

func NewMessageQueueFixture(queue queues.IMessageQueue) *MessageQueueFixture {
	c := MessageQueueFixture{
		queue: queue,
	}
	return &c
}

func (c *MessageQueueFixture) TestSendReceiveMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sndErr)

	time.Sleep(100 * time.Millisecond)

	// if c.queue.GetCapabilities().CanMessageCount() {
	// 	count, rdErr := c.queue.MessageCount()
	// 	assert.Nil(t, rdErr)
	// 	assert.Greater(t, count, (int64)(0))
	// }

	envelope2, rcvErr := c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)
}

func (c *MessageQueueFixture) TestMessageCount(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	c.queue.Send(context.Background(), "", envelope1)

	time.Sleep(500 * time.Millisecond)

	count, err := c.queue.ReadMessageCount()
	assert.Nil(t, err)
	assert.True(t, count >= 1)
}

func (c *MessageQueueFixture) TestReceiveSendMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))

	time.AfterFunc(500*time.Millisecond, func() {
		c.queue.Send(context.Background(), "", envelope1)
	})

	envelope2, rcvErr := c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)
}

func (c *MessageQueueFixture) TestReceiveCompleteMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sndErr)

	time.Sleep(100 * time.Millisecond)

	// count, rdErr := c.queue.ReadMessageCount()
	// assert.Nil(t, rdErr)
	// assert.Greater(t, count, (int64)(0))

	envelope2, rcvErr := c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	cplErr := c.queue.Complete(context.Background(), envelope2)
	assert.Nil(t, cplErr)
	assert.Nil(t, envelope2.GetReference())
}

func (c *MessageQueueFixture) TestReceiveAbandonMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sndErr)

	envelope2, rcvErr := c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	abdErr := c.queue.Abandon(context.Background(), envelope2)
	assert.Nil(t, abdErr)

	envelope2, rcvErr = c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)
}

func (c *MessageQueueFixture) TestSendPeekMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sndErr)

	time.Sleep(100 * time.Millisecond)

	envelope2, pkErr := c.queue.Peek(context.Background(), "")
	assert.Nil(t, pkErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	// pop message from queue for next test
	_, rcvErr := c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
}

func (c *MessageQueueFixture) TestPeekNoMessage(t *testing.T) {
	envelope, pkErr := c.queue.Peek(context.Background(), "")
	assert.Nil(t, pkErr)
	assert.Nil(t, envelope)
}

func (c *MessageQueueFixture) TestMoveToDeadMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sndErr)

	envelope2, rcvErr := c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	mvErr := c.queue.MoveToDeadLetter(context.Background(), envelope2)
	assert.Nil(t, mvErr)
}

func (c *MessageQueueFixture) TestOnMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	receiver := &TestMsgReceiver{}
	c.queue.BeginListen(context.Background(), "", receiver)

	time.Sleep(1000 * time.Millisecond)

	sndErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sndErr)

	time.Sleep(1000 * time.Millisecond)

	envelope2 := receiver.GetEnvelope()
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	c.queue.EndListen(context.Background(), "")
}

type TestMsgReceiver struct {
	lock      sync.Mutex
	_envelope *queues.MessageEnvelope
}

func (c *TestMsgReceiver) GetEnvelope() queues.MessageEnvelope {
	c.lock.Lock()
	defer c.lock.Unlock()
	return *c._envelope
}

func (c *TestMsgReceiver) ReceiveMessage(ctx context.Context, envelope *queues.MessageEnvelope, queue queues.IMessageQueue) (err error) {
	c.lock.Lock()
	defer c.lock.Unlock()
	c._envelope = envelope
	return nil
}


test/connect/NatsConnection_test.go
package test_connect

import (
	"context"
	"os"
	"testing"

	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	connect "github.com/pip-services3-gox/pip-services3-nats-gox/connect"
	"github.com/stretchr/testify/assert"
)

type natsConnectionTest struct {
	connection *connect.NatsConnection
}

func newNatsConnectionTest() *natsConnectionTest {
	natsUri := os.Getenv("NATS_SERVICE_URI")
	natsHost := os.Getenv("NATS_SERVICE_HOST")
	if natsHost == "" {
		natsHost = "localhost"
	}

	natsPort := os.Getenv("NATS_SERVICE_PORT")
	if natsPort == "" {
		natsPort = "4222"
	}

	natsToken := os.Getenv("NATS_TOKEN")
	// if natsToken == "" {
	// 	natsToken = ""
	// }
	natsUser := os.Getenv("NATS_USER")
	// if natsUser == "" {
	// 	natsUser = ""
	// }
	natsPassword := os.Getenv("NATS_PASS")
	// if natsPassword == "" {
	// 	natsPassword = ""
	// }

	if natsUri == "" && natsHost == "" {
		return nil
	}

	connection := connect.NewNatsConnection()
	connection.Configure(context.Background(), cconf.NewConfigParamsFromTuples(
		"connection.uri", natsUri,
		"connection.host", natsHost,
		"connection.port", natsPort,
		"credential.token", natsToken,
		"credential.username", natsUser,
		"credential.password", natsPassword,
	))

	return &natsConnectionTest{
		connection: connection,
	}
}

func (c *natsConnectionTest) TestOpenClose(t *testing.T) {
	ctx := context.Background()

	err := c.connection.Open(ctx, "")
	assert.Nil(t, err)
	assert.True(t, c.connection.IsOpen())
	assert.NotNil(t, c.connection.GetConnection())

	err = c.connection.Close(ctx, "")
	assert.Nil(t, err)
	assert.False(t, c.connection.IsOpen())
	assert.Nil(t, c.connection.GetConnection())
}

func TestNatsConnection(t *testing.T) {
	c := newNatsConnectionTest()
	if c == nil {
		return
	}

	t.Run("Open and Close", c.TestOpenClose)
}


test/queues/NatsMessageQueue_test.go
package test_queues

import (
	"context"
	"os"
	"testing"

	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	queues "github.com/pip-services3-gox/pip-services3-nats-gox/queues"
)

type natsMessageQueueTest struct {
	queue   *queues.NatsMessageQueue
	fixture *MessageQueueFixture
}

func newNatsMessageQueueTest() *natsMessageQueueTest {
	natsUri := os.Getenv("NATS_SERVICE_URI")
	natsHost := os.Getenv("NATS_SERVICE_HOST")
	if natsHost == "" {
		natsHost = "localhost"
	}

	natsPort := os.Getenv("NATS_SERVICE_PORT")
	if natsPort == "" {
		natsPort = "4222"
	}

	natsQueue := os.Getenv("NATS_QUEUE")
	if natsQueue == "" {
		natsQueue = "test"
	}

	natsToken := os.Getenv("NATS_TOKEN")
	if natsToken == "" {
		natsToken = ""
	}

	natsUser := os.Getenv("NATS_USER")
	if natsUser == "" {
		natsUser = "nats"
	}
	natsPassword := os.Getenv("NATS_PASS")
	if natsPassword == "" {
		natsPassword = "nats"
	}

	if natsUri == "" && natsHost == "" {
		return nil
	}

	queue := queues.NewNatsMessageQueue(natsQueue)
	queue.Configure(context.Background(), cconf.NewConfigParamsFromTuples(
		"connection.uri", natsUri,
		"connection.host", natsHost,
		"connection.port", natsPort,
		"credential.token", natsToken,
		"credential.username", natsUser,
		"credential.password", natsPassword,
		"options.autosubscribe", true,
	))

	fixture := NewMessageQueueFixture(queue)

	return &natsMessageQueueTest{
		queue:   queue,
		fixture: fixture,
	}
}

func (c *natsMessageQueueTest) setup(t *testing.T) {
	err := c.queue.Open(context.Background(), "")
	if err != nil {
		t.Error("Failed to open queue", err)
		return
	}

	// err = c.queue.Clear("")
	// if err != nil {
	// 	t.Error("Failed to clear queue", err)
	// 	return
	// }
}

func (c *natsMessageQueueTest) teardown(t *testing.T) {
	err := c.queue.Close(context.Background(), "")
	if err != nil {
		t.Error("Failed to close queue", err)
	}
}

func TestNatsMessageQueue(t *testing.T) {
	c := newNatsMessageQueueTest()
	if c == nil {
		return
	}

	c.setup(t)
	t.Run("Send Receive Message", c.fixture.TestSendReceiveMessage)
	c.teardown(t)

	c.setup(t)
	t.Run("Receive Send Message", c.fixture.TestReceiveSendMessage)
	c.teardown(t)

	c.setup(t)
	t.Run("Send Peek Message", c.fixture.TestSendPeekMessage)
	c.teardown(t)

	c.setup(t)
	t.Run("Peek No Message", c.fixture.TestPeekNoMessage)
	c.teardown(t)

	c.setup(t)
	t.Run("On Message", c.fixture.TestOnMessage)
	c.teardown(t)
}


connect/NatsConnectionResolver.go
package connect

import (
	"context"
	"strconv"
	"strings"

	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	cerr "github.com/pip-services3-gox/pip-services3-commons-gox/errors"
	cref "github.com/pip-services3-gox/pip-services3-commons-gox/refer"
	cauth "github.com/pip-services3-gox/pip-services3-components-gox/auth"
	ccon "github.com/pip-services3-gox/pip-services3-components-gox/connect"
)

// NatsConnectionResolver helper class that resolves Nats connection and credential parameters,
// validates them and generates connection options.
//
//	Configuration parameters:
//
//	- connection(s):
//		- discovery_key:               (optional) a key to retrieve the connection from IDiscovery
//		- host:                        host name or IP address
//		- port:                        port number
//		- uri:                         resource URI or connection string with all parameters in it
//	- credential(s):
//		- store_key:                   (optional) a key to retrieve the credentials from ICredentialStore
//		- username:                    user name
//		- password:                    user password
//
//	References:
//
//		- *:discovery:*:*:1.0          (optional) IDiscovery services to resolve connections
//		- *:credential-store:*:*:1.0   (optional) Credential stores to resolve credentials

type NatsConnectionResolver struct {
	// The connections resolver.
	ConnectionResolver *ccon.ConnectionResolver
	//The credentials resolver.
	CredentialResolver *cauth.CredentialResolver
}

func NewNatsConnectionResolver() *NatsConnectionResolver {
	c := NatsConnectionResolver{}
	c.ConnectionResolver = ccon.NewEmptyConnectionResolver()
	c.CredentialResolver = cauth.NewEmptyCredentialResolver()
	return &c
}

// Configure are configures component by passing configuration parameters.
// Parameters:
//  - config   *cconf.ConfigParams
// configuration parameters to be set.
func (c *NatsConnectionResolver) Configure(ctx context.Context, config *cconf.ConfigParams) {
	c.ConnectionResolver.Configure(ctx, config)
	c.CredentialResolver.Configure(ctx, config)
}

// SetReferences are sets references to dependent components.
// Parameters:
//  - references  cref.IReferences
//	references to locate the component dependencies.
func (c *NatsConnectionResolver) SetReferences(ctx context.Context, references cref.IReferences) {
	c.ConnectionResolver.SetReferences(ctx, references)
	c.CredentialResolver.SetReferences(ctx, references)
}

func (c *NatsConnectionResolver) validateConnection(correlationId string, connection *ccon.ConnectionParams) error {
	if connection == nil {
		return cerr.NewConfigError(correlationId, "NO_CONNECTION", "Nats connection is not set")
	}

	uri := connection.Uri()
	if uri != "" {
		return nil
	}

	protocol := connection.ProtocolWithDefault("nats")
	if protocol == "" {
		return cerr.NewConfigError(correlationId, "NO_PROTOCOL", "Connection protocol is not set")
	}
	if protocol != "nats" {
		return cerr.NewConfigError(correlationId, "UNSUPPORTED_PROTOCOL", "The protocol "+protocol+" is not supported")
	}

	host := connection.Host()
	if host == "" {
		return cerr.NewConfigError(correlationId, "NO_HOST", "Connection host is not set")
	}

	port := connection.PortWithDefault(4222)
	if port == 0 {
		return cerr.NewConfigError(correlationId, "NO_PORT", "Connection port is not set")
	}

	return nil
}

func (c *NatsConnectionResolver) composeOptions(connections []*ccon.ConnectionParams,
	credential *cauth.CredentialParams) *cconf.ConfigParams {

	// Define additional parameters parameters
	if credential == nil {
		credential = cauth.NewEmptyCredentialParams()
	}

	// Contruct options and copy over credentials
	options := cconf.NewEmptyConfigParams().SetDefaults(credential.ConfigParams)

	globalUri := ""
	uriBuilder := strings.Builder{}

	// Process connections, find or constract uri
	for _, connection := range connections {
		options = options.SetDefaults(connection.ConfigParams)

		if globalUri != "" {
			continue
		}

		uri := connection.Uri()
		if uri != "" {
			globalUri = uri
			continue
		}

		if uriBuilder.Len() > 0 {
			uriBuilder.WriteString(", ")
		}

		protocol := connection.ProtocolWithDefault("nats")
		uriBuilder.WriteString(protocol)

		host := connection.Host()
		uriBuilder.WriteString("://")
		uriBuilder.WriteString(host)

		port := connection.PortWithDefault(4222)
		uriBuilder.WriteString(":")
		uriBuilder.WriteString(strconv.Itoa(port))
	}

	// Set connection uri
	if globalUri != "" {
		options.SetAsObject("uri", globalUri)
	} else {
		options.SetAsObject("uri", uriBuilder.String())
	}

	return options
}

// Resolves Nats connection options from connection and credential parameters.
// Parameters:
//   - correlationId   string
//   (optional) transaction id to trace execution through call chain.
// Returns options *cconf.ConfigParams, err error
// receives resolved options or error.
func (c *NatsConnectionResolver) Resolve(correlationId string) (*cconf.ConfigParams, error) {
	connections, err := c.ConnectionResolver.ResolveAll(correlationId)
	if err != nil {
		return nil, err
	}

	credential, err := c.CredentialResolver.Lookup(context.Background(), correlationId)
	if err != nil {
		return nil, err
	}

	// Validate connections
	for _, connection := range connections {
		err = c.validateConnection(correlationId, connection)
		if err != nil {
			return nil, err
		}
	}

	options := c.composeOptions(connections, credential)
	return options, nil
}

// Compose method are composes Nats connection options from connection and credential parameters.
// Parameters:
//   - correlationId  string  (optional) transaction id to trace execution through call chain.
//   - connection  *ccon.ConnectionParams    connection parameters
//   - credential  *cauth.CredentialParams   credential parameters
// Returns: options *cconf.ConfigParams, err error
// resolved options or error.
func (c *NatsConnectionResolver) Compose(correlationId string, connections []*ccon.ConnectionParams, credential *cauth.CredentialParams) (*cconf.ConfigParams, error) {
	// Validate connections
	for _, connection := range connections {
		err := c.validateConnection(correlationId, connection)
		if err != nil {
			return nil, err
		}
	}

	options := c.composeOptions(connections, credential)
	return options, nil
}


queues/NatsMessageQueue.go
package queues

import (
	"context"
	"fmt"
	"time"

	"github.com/nats-io/nats.go"
	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	cqueues "github.com/pip-services3-gox/pip-services3-messaging-gox/queues"
)

//NatsMessageQueue are message queue that sends and receives messages via NATS message broker.
//
//	Configuration parameters:
//
//		- subject:                       name of NATS topic (subject) to subscribe
//		- queue_group:                   name of NATS queue group
//		- connection(s):
//			- discovery_key:               (optional) a key to retrieve the connection from  IDiscovery
//			- host:                        host name or IP address
//			- port:                        port number
//			- uri:                         resource URI or connection string with all parameters in it
//		- credential(s):
//			- store_key:                   (optional) a key to retrieve the credentials from  ICredentialStore
//			- username:                    user name
//			- password:                    user password
//		- options:
//			- serialize_message:    (optional) true to serialize entire message as JSON, false to send only message payload (default: true)
//			- autosubscribe:        (optional) true to automatically subscribe on option (default: false)
//			- retry_connect:        (optional) turns on/off automated reconnect when connection is log (default: true)
//			- max_reconnect:        (optional) maximum reconnection attempts (default: 3)
//			- reconnect_timeout:    (optional) number of milliseconds to wait on each reconnection attempt (default: 3000)
//			- flush_timeout:        (optional) number of milliseconds to wait on flushing messages (default: 3000)
//
//
// References:
//
//		- *:logger:*:*:1.0             (optional)  ILogger components to pass log messages
//		- *:counters:*:*:1.0           (optional)  ICounters components to pass collected measurements
//		- *:discovery:*:*:1.0          (optional)  IDiscovery services to resolve connections
//		- *:credential-store:*:*:1.0   (optional) Credential stores to resolve credentials
//		- *:connection:nats:*:1.0      (optional) Shared connection to NATS service
//
// See MessageQueue
// See MessagingCapabilities
//
//	Example:
//		ctx := context.Background()
//		queue := NewNatsMessageQueue("myqueue")
//		queue.Configure(ctx, cconf.NewConfigParamsFromTuples(
//			"subject", "mytopic",
//			"queue_group", "mygroup",
//			"connection.protocol", "nats"
//			"connection.host", "localhost"
//			"connection.port", 1883
//		))
//
//		_ = queue.Open(ctx, "123")
//
//		_ = queue.Send(ctx, "123", NewMessageEnvelope("", "mymessage", "ABC"))
//
//		message, err := queue.Receive(ctx, "123", 10000*time.Milliseconds)
//		if (message != nil) {
//			...
//			queue.Complete(ctx, message);
//		}

type NatsMessageQueue struct {
	*NatsAbstractMessageQueue

	autoSubscribe bool
	subscribed    bool
	messages      []*cqueues.MessageEnvelope
	receiver      cqueues.IMessageReceiver
}

// NewNatsMessageQueue are creates a new instance of the message queue.
//	Parameters:
//		- ctx context.Context	operation context
//		- name  string (optional) a queue name.
func NewNatsMessageQueue(name string) *NatsMessageQueue {
	c := NatsMessageQueue{}

	c.NatsAbstractMessageQueue = InheritNatsAbstractMessageQueue(&c, name,
		cqueues.NewMessagingCapabilities(false, true, true, true, true, false, false, false, true))

	c.messages = make([]*cqueues.MessageEnvelope, 0)

	return &c
}

// Configures component by passing configuration parameters.
//	Parameters:
//		- ctx context.Context	operation context
//		- config    configuration parameters to be set.
func (c *NatsMessageQueue) Configure(ctx context.Context, config *cconf.ConfigParams) {
	c.NatsAbstractMessageQueue.Configure(ctx, config)

	c.autoSubscribe = config.GetAsBooleanWithDefault("options.autosubscribe", c.autoSubscribe)
}

// Opens the component with given connection and credential parameters.
//	Parameters:
//		- ctx context.Context	operation context
//		- correlationId     (optional) transaction id to trace execution through call chain.
// Returns error or nil no errors occured.
func (c *NatsMessageQueue) Open(ctx context.Context, correlationId string) error {
	if c.IsOpen() {
		return nil
	}

	err := c.NatsAbstractMessageQueue.Open(ctx, correlationId)
	if err != nil {
		return err
	}

	// Subscribe right away
	if c.autoSubscribe {
		err = c.subscribe(ctx, correlationId)
		if err != nil {
			c.Close(ctx, correlationId)
			return err
		}
	}

	return nil
}

// Close method are Closes component and frees used resources.
//	Parameters:
//		- ctx context.Context	operation context
//		- correlationId string 	(optional) transaction id to trace execution through call chain.
// Returns error or nil no errors occured.
func (c *NatsMessageQueue) Close(ctx context.Context, correlationId string) error {
	if !c.IsOpen() {
		return nil
	}

	err := c.NatsAbstractMessageQueue.Close(ctx, correlationId)

	// Unsubscribe from topic
	if c.subscribed {
		subject := c.SubscriptionSubject()
		c.Connection.Unsubscribe(ctx, subject, c.QueueGroup, c)
		c.subscribed = false
	}

	c.Lock.Lock()
	defer c.Lock.Unlock()
	c.receiver = nil
	c.messages = make([]*cqueues.MessageEnvelope, 0)

	return err
}

func (c *NatsMessageQueue) subscribe(ctx context.Context, correlationId string) error {
	c.Lock.Lock()
	defer c.Lock.Unlock()

	// Check if already were subscribed
	if c.subscribed {
		return nil
	}

	// Subscribe to the topic
	subject := c.SubscriptionSubject()
	err := c.Connection.Subscribe(ctx, subject, c.QueueGroup, c)
	if err != nil {
		c.Logger.Error(ctx, correlationId, err, "Failed to subscribe to subject "+subject)
		return err
	}

	c.subscribed = true
	return nil
}

// Clear method are clears component state.
//	Parameters:
//		- ctx context.Context	operation context
//		- correlationId 	string (optional) transaction id to trace execution through call chain.
// Returns error or nil no errors occured.
func (c *NatsMessageQueue) Clear(ctx context.Context, correlationId string) (err error) {
	c.Lock.Lock()
	defer c.Lock.Unlock()

	c.messages = make([]*cqueues.MessageEnvelope, 0)
	c.receiver = nil

	return nil
}

// ReadMessageCount method are reads the current number of messages in the queue to be delivered.
// Returns number of messages or error.
func (c *NatsMessageQueue) ReadMessageCount() (count int64, err error) {
	c.Lock.Lock()
	defer c.Lock.Unlock()

	count = (int64)(len(c.messages))
	return count, nil
}

// Peek method are peeks a single incoming message from the queue without removing it.
// If there are no messages available in the queue it returns nil.
//	Parameters:
//		- ctx context.Context	operation context
//		- correlationId  string  (optional) transaction id to trace execution through call chain.
// Returns: result *cqueues.MessageEnvelope, err error
// message or error.
func (c *NatsMessageQueue) Peek(ctx context.Context, correlationId string) (*cqueues.MessageEnvelope, error) {
	err := c.CheckOpen(correlationId)
	if err != nil {
		return nil, err
	}

	// Subscribe if needed
	err = c.subscribe(ctx, correlationId)
	if err != nil {
		return nil, err
	}

	var message *cqueues.MessageEnvelope

	// Pick a message
	c.Lock.Lock()
	if len(c.messages) > 0 {
		message = c.messages[0]
	}
	c.Lock.Unlock()

	if message != nil {
		c.Logger.Trace(ctx, message.CorrelationId, "Peeked message %s on %s", message, c.String())
	}

	return message, nil
}

// PeekBatch method are peeks multiple incoming messages from the queue without removing them.
// If there are no messages available in the queue it returns an empty list.
// Important: This method is not supported by NATS.
//	Parameters:
//		- ctx context.Context	operation context
//		- correlationId     (optional) transaction id to trace execution through call chain.
//		- messageCount      a maximum number of messages to peek.
// Returns:          receives a list with messages or error.
func (c *NatsMessageQueue) PeekBatch(ctx context.Context, correlationId string, messageCount int64) ([]*cqueues.MessageEnvelope, error) {
	err := c.CheckOpen(correlationId)
	if err != nil {
		return nil, err
	}

	// Subscribe if needed
	err = c.subscribe(ctx, correlationId)
	if err != nil {
		return nil, err
	}

	messages := make([]*cqueues.MessageEnvelope, messageCount)

	c.Lock.Lock()
	if messageCount <= (int64)(len(messages)) {
		copy(messages, c.messages[0:messageCount])
	}
	c.Lock.Unlock()

	c.Logger.Trace(ctx, correlationId, "Peeked %d messages on %s", len(messages), c.Name())

	return messages, nil
}

// Receive method are receives an incoming message and removes it from the queue.
//	Parameters:
//		- ctx context.Context	operation context
//		- correlationId   string   (optional) transaction id to trace execution through call chain.
//		- waitTimeout  time.Duration     a timeout in milliseconds to wait for a message to come.
// Returns:  result *cqueues.MessageEnvelope, err error
// receives a message or error.
func (c *NatsMessageQueue) Receive(ctx context.Context, correlationId string, waitTimeout time.Duration) (*cqueues.MessageEnvelope, error) {
	err := c.CheckOpen(correlationId)
	if err != nil {
		return nil, err
	}

	// Subscribe if needed
	err = c.subscribe(ctx, correlationId)
	if err != nil {
		return nil, err
	}

	messageReceived := false
	var message *cqueues.MessageEnvelope
	elapsedTime := time.Duration(0)

	for elapsedTime < waitTimeout && !messageReceived {
		c.Lock.Lock()
		if len(c.messages) == 0 {
			c.Lock.Unlock()
			time.Sleep(time.Duration(100) * time.Millisecond)
			elapsedTime += time.Duration(100)
			continue
		}

		// Get message from the queue
		message = c.messages[0]
		c.messages = c.messages[1:]

		// Add messages to locked messages list
		messageReceived = true
		c.Lock.Unlock()
	}

	return message, nil
}

// Function thath process incoming messages
//	Parameters:
//		- msg *nats.Msg	message from the NATs
func (c *NatsMessageQueue) OnMessage(msg *nats.Msg) {
	// Deserialize message
	message, err := c.ToMessage(msg)
	if err != nil {
		c.Logger.Error(context.Background(), "", err, "Failed to read received message")
	}

	c.Counters.IncrementOne(context.Background(), "queue."+c.Name()+".received_messages")
	c.Logger.Debug(context.Background(), message.CorrelationId, "Received message %s via %s", msg, c.Name())

	// Send message to receiver if its set or put it into the queue
	c.Lock.Lock()
	if c.receiver != nil {
		receiver := c.receiver
		c.Lock.Unlock()
		c.sendMessageToReceiver(receiver, message)
	} else {
		c.messages = append(c.messages, message)
		c.Lock.Unlock()
	}
}

func (c *NatsMessageQueue) sendMessageToReceiver(receiver cqueues.IMessageReceiver, message *cqueues.MessageEnvelope) {
	correlationId := message.CorrelationId

	defer func() {
		if r := recover(); r != nil {
			err := fmt.Sprintf("%v", r)
			c.Logger.Error(context.Background(), correlationId, nil, "Failed to process the message - "+err)
		}
	}()

	err := receiver.ReceiveMessage(context.Background(), message, c)
	if err != nil {
		c.Logger.Error(context.Background(), correlationId, err, "Failed to process the message")
	}
}

// Listens for incoming messages and blocks the current thread until queue is closed.
//	Parameters:
//		- ctx context.Context	operation context
//		- correlationId   string  (optional) transaction id to trace execution through call chain.
//		- receiver    cqueues.IMessageReceiver      a receiver to receive incoming messages.
//
// See IMessageReceiver
// See receive
func (c *NatsMessageQueue) Listen(ctx context.Context, correlationId string, receiver cqueues.IMessageReceiver) error {
	err := c.CheckOpen(correlationId)
	if err != nil {
		return err
	}

	// Subscribe if needed
	err = c.subscribe(ctx, correlationId)
	if err != nil {
		return err
	}

	c.Logger.Trace(ctx, "", "Started listening messages at %s", c.Name())

	// Get all collected messages
	c.Lock.Lock()
	batchMessages := c.messages
	c.messages = []*cqueues.MessageEnvelope{}
	c.Lock.Unlock()

	// Resend collected messages to receiver
	for _, message := range batchMessages {
		receiver.ReceiveMessage(ctx, message, c)
	}

	// Set the receiver
	c.Lock.Lock()
	c.receiver = receiver
	c.Lock.Unlock()

	return nil
}

// EndListen method are ends listening for incoming messages.
// When this method is call listen unblocks the thread and execution continues.
//	Parameters:
//		- ctx context.Context	operation context
//		- correlationId  string   (optional) transaction id to trace execution through call chain.
func (c *NatsMessageQueue) EndListen(ctx context.Context, correlationId string) {
	c.Lock.Lock()
	c.receiver = nil
	c.Lock.Unlock()
}
